### PR TITLE
add RBAC support to proxyless gRPC

### PR DESCRIPTION
**Pull request by costinm**
Generate RBAC HTTP filters for proxyless gRPC listeneners.

Important: gRPC supports a subset of Istio attributes and capabilities. This is experimental - we still don't have a good 
solution for what happens if an unsupported attribute is used. Also due to a bug, current gRPC version rejects configs
with both ALLOW and DENY filters - will be fixed on their side. 

Using the local test - RBAC generation doesn't seem to need to run in k8s, behaves the same way. Also added mTLS 
testing to the local test.

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Looks good. Will let @stevenctl approve as they are the expert

File: pilot/pkg/networking/grpcgen/lds.go [hunk 1]
```diff
@@ -185,6 +222,40 @@ func buildInboundFilterChain(nameSuffix string, tlsContext *tls.DownstreamTlsCon
...
+// TODO: gRPC also supports 'per route override' - not yet clear how to use it, Istio uses path expressions instead and we don't generate
+// vhosts or routes for the inbound listener.
+//
+// For gateways it would make a lot of sense to use this concept, same for moving path prefix at top level ( more scalable, easier for users)
+// This should probably be done for the v2 API.
+//
+//
+//
+// nolint: unparam
+func buildRBAC(node *model.Proxy, push *model.PushContext, suffix string, context *tls.DownstreamTlsContext, a rbacpb.RBAC_Action, policies []model.AuthorizationPolicy) *rbacpb.RBAC {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/grpcgen/lds.go:239]: Can we reuse the authz builder, in case there are any changes in the future there?

Comment (costinm) [reply pilot/pkg/networking/grpcgen/lds.go:239]: I don't think so, like LDS the generated output is slightly different.

However in future I am trying to define a subset of RBAC that can be used for both Envoy and proxyless, and works as  K8S Gateway policy attachment at route level ( supported by Envoy, not by current Istio API). The ideal is for this new API to be shared across all proxies ( and hopefully adopted by other gateway vendors ), and to work better with gateways ( O(1) vs O(N), etc).

File: pilot/pkg/networking/grpcgen/grpcgen_test.go [hunk 2]
```diff
@@ -51,103 +68,83 @@ var (
...
+					InstanceIPs: []string{""},
 				},
 			},
 		},
 		DiscoveryAddress: grpcXdsAddr,
 		CertDir:          path.Join(env.IstioSrc, "tests/testdata/certs/default"),
 	})
 	if err != nil {
-		return nil, fmt.Errorf("failed generating bootstrap: %v", err)
+		return []byte{}
```
[the remark this answers is not in the document]
Comment (costinm) [reply pilot/pkg/networking/grpcgen/grpcgen_test.go:106]: It's not hard to debug - the bootstrap is logged, and it's more convenient. There are not many common errors in bootstrap generation.

File: pilot/pkg/networking/grpcgen/grpcgen_test.go [hunk 3]
```diff
@@ -51,103 +68,83 @@ var (
...
 func TestGRPC(t *testing.T) {
-	xdsresolver := resolverForTest(t, "istio-system")
-	ds := xds.NewXDS(make(chan struct{}))
+	// Istio already has a dependency on Zap and istio-ecosystem - set the logger
+	// so we can explicitly enable level/verbosity. Alternative is to use env variables.
+	// '4' skips the adapters, shows the actual code logging the info.
+	// 3  works for most core - but fails for 'prefixLogger' - used by XDS
+	zl, _ := zap.NewDevelopment(zap.AddCallerSkip(4))
+	// TODO: use zap.observer, we can use the collected info for assertions.
+	grpc_zap.ReplaceGrpcLoggerV2WithVerbosity(zl, 99)
```
[the remark this answers is not in the document]
Comment (costinm) [reply pilot/pkg/networking/grpcgen/grpcgen_test.go]: It was more of an experiment on how to set grpc logging for real apps ( where istio log won't be used ).

Reverted, I see you added an init()

File: pilot/pkg/networking/grpcgen/grpcgen_test.go [hunk 4]
```diff
@@ -213,8 +250,175 @@ func TestGRPC(t *testing.T) {
...
+		t.Fatalf("net.Listen(tcp, %q) failed: %v", port, err)
+	}
+
+	go func() {
+		err := grpcServer.Serve(lis)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}()
+	time.Sleep(3 * time.Second)
```
[the remark this answers is not in the document]
Comment (stevenctl) [reply pilot/pkg/networking/grpcgen/grpcgen_test.go:403]: +1

**Review by stevenctl — APPROVED** [review]
a couple of nits on tests, either you or I can fix them in a follow-up.

File: pilot/pkg/networking/grpcgen/grpcgen_test.go [hunk 5]
```diff
@@ -192,21 +189,37 @@ func TestGRPC(t *testing.T) {
...
+		})
 	})
 
 	t.Run("gRPC-dial", func(t *testing.T) {
 		for _, host := range []string{
 			"istiod.istio-system.svc.cluster.local",
-			"istiod.istio-system.svc",
-			"istiod.istio-system",
-			"istiod",
+			//"istiod.istio-system.svc",
```
[the remark this answers is not in the document]
Comment (costinm) [reply pilot/pkg/networking/grpcgen/grpcgen_test.go:217]: Right, will fix it in a separate PR. Except the last one - I moved the test to be in a different namespace.

**Pull request merged**